### PR TITLE
[Experimental] Use variations to provide all the product filters wrapped by CollectionFilters block

### LINKED ISSUE #42160: Inserter behavior for collection filter blocks
New filter blocks are inner block of the Collection Filters block, this is our deal-breaker. So by design, new filter blocks can't be added everywhere -> this can make merchants confused:
- Current filter blocks can be added every where.
- Most merchants will search for 'price' when they want to insert the price filter block, not `collection filters`.

We need a solution to maintain the current discoverability for new filter blocks.

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

To support discoverability and simplify adding filters this PR registers each of the new interactivity filters as a variation of the collection-filters block and inside the block we choose a template for the respective filter based on an attribute passed in the variant.

Closes https://github.com/woocommerce/woocommerce/issues/42160

### How to test the changes in this Pull Request:

1. In the block editor the filter blocks by themselves should not show in the block inserter
2. These new block variants should show
3. try adding each one of the new variants
4. It should add a collection filters wrapper and a corresponding heading for the block

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
[Experimental] Use variations of collection filter to provide new interactivity based filter blocks.

#### Comment 

</details>

### CODE REVIEW

**Review by dinhtungdu — COMMENTED** [review]
Thanks for working on this PR! This is working as expected according to the testing instructions. I don't have any comment codewise but about one on the behavior:

> In the block editor the filter blocks by themselves should not show in the block inserter

I don't think we should do this, because:
- The filter blocks are only available inside `Collection Filters` block, so they don't add more noise to the global inserter.
- With this PR, when adding inner blocks to `Collection Filters`, for example Price Filter, we're adding another `Collection Filters` block and `Price Filter` blocks to the original `Collection Filters` block.

I think the behavior should be:
- `Collection Filters` block shouldn't be available inside other `Collection Filters` block.
- The new variations added in this PR also shouldn't be available inside `Collection Filters` block.
- The inserter visibility of new filter blocks should be restored as they are before, only available inside `Collection Filters` block.
- In the other words, we should disallow the following blocks inside `Collection Filters`, so when inside that block, the only result, price filter for example, should be the new price filter block:
  - Filter Wrapper variations (we will disable legacy blocks soon so this may not necessary).
  - Collection Filters block and its variations.

Sadly, Gutenberg [hasn't provided any API to disallow specific inner blocks](https://github.com/WordPress/gutenberg/issues/41236#:~:text=Offer%20a%20way%20to%20define,the%20currently%20supported%20allowed%20list) right now, so we have to either:
- use a fix list of allowed blocks, which includes new filter blocks and blocks for layout purpose.
- use a dynamic list of allow blocks (defining the list at run time).

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/block.json [hunk 1]
```diff
@@ -4,26 +4,23 @@
...
-	],
+	"usesContext": [ "query", "queryId" ],
 	"providesContext": {
 		"collectionData": "collectionData"
 	},
 	"attributes": {
 		"collectionData": {
 			"type": "object",
 			"default": {}
+		},
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/block.json:21]: some autoformatting left behind from editing this file. happy to remove this from PR if needed.

**Comment by samueljseay** [comment]
@dinhtungdu I took another go at this based on your feedback. Hurts the head a little when dealing with variants and such but let me try explain the behavior implemented. If you could test and see if what I've implemented here matches your expectations :)

1. Collection filters can be inserted. It has default variant which includes all the new filter blocks

2. Collection filters cannot be inserted in collection filters

3. The filters themselves can still be inserted at top level of collection filters though

4. The filter variants cannot be nested in filter variants (again avoiding nested collection filters)

5. The inserter false is removed now and you can use variants to add the collection-filter -> filter hierarchy

6. By disallowing woocommerce/filter-wrapper we disallow adding any of the old filters


I think that about sums it up.

I expect we can fix up any block naming as part of https://github.com/woocommerce/woocommerce/issues/43167

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/block.json (hunk 1 above)
Comment (dinhtungdu) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/block.json:21]: No issue, looks like it's me formatted these block.json before 😅

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx [hunk 2]
```diff
@@ -1,35 +1,93 @@
...
+	// don't allow nesting variants
+	'woocommerce/collection-filters-variant',
+	'woocommerce/rating-filter-variant',
+	'woocommerce/price-filter-variant',
+	'woocommerce/active-filters-variant',
+	'woocommerce/attribute-filter-variant',
+	'woocommerce/stock-filter-variant',
+	// don't allow nesting old filter blocks
+	'woocommerce/filter-wrapper',
+];
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]:
```suggestion
const DISALLOWED_BLOCKS = [
	'woocommerce/filter-wrapper',
	'woocommerce/collection-filters',
];
```

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx [hunk 3]
```diff
@@ -1,35 +1,93 @@
...
-	const innerBlockProps = useInnerBlocksProps( blockProps, {
-		template,
-	} );
+const Edit = ( props: EditProps ) => {
+	const filterType = props.attributes.filterType;
+
+	const allowedBlocks =
+		filterType === 'collection-filters'
+			? getAllowedBlocks( DISALLOWED_COLLECTION_FILTERS_BLOCKS )
+			: getAllowedBlocks( DISALLOWED_BLOCKS );
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]:
```suggestion
	const allowedBlocks = getAllowedBlocks( DISALLOWED_BLOCKS );
```

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/index.tsx [hunk 4]
```diff
@@ -22,6 +31,124 @@ if ( isExperimentalBuild() ) {
 				/>
 			),
 		},
+		variations: [
+			{
+				name: 'woocommerce/collection-filters-variant',
+				title: __( 'Collection Filters (v)', 'woocommerce' ),
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/index.tsx]: I guess the `(v)` is for debugging purpose, right?

**Review by dinhtungdu — COMMENTED** [review]
Great update! Having the Collection Filters inserts all filter blocks by default is a good UX 💯 .

In my test, I can still insert variation inside another variation. Here is the testing video taken on https://shy-kingfisher.jurassic.ninja/

https://github.com/woocommerce/woocommerce/assets/5423135/543e4a7a-d087-4cde-b498-528ea65c4c43

Checking the code, I can see you're splitting into two cases. I don't think we need those two cases because in every case, only the collection filter blocks can be added in inside `Collection Filters`, not variations of `Collection Filters` block.

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/index.tsx (hunk 4 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/index.tsx]: Oops yes left this in by accident. Thanks for spotting.

Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/index.tsx]: Resolved in https://github.com/woocommerce/woocommerce/pull/43216/commits/1692a4ef3381f442fb7dc7ab90977cf836753e8a

**Comment by samueljseay** [comment]
thanks @dinhtungdu I tested your changes and you're right I was overthinking the block filtering (I also learned that variations are not individual block types, neat!)

I've applied your suggestions and ready for re-review

**Review by dinhtungdu — APPROVED** [review]
Thanks for the updates! This is LGTM now.

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx [hunk 5]
```diff
@@ -1,35 +1,62 @@
...
-		template,
-	} );
+const Edit = ( props: EditProps ) => {
+	const allowedBlocks = getAllowedBlocks( DISALLOWED_BLOCKS );
+
+	const template = templates[ props.attributes.filterType ];
 
-	return <nav { ...innerBlockProps } />;
+	return (
+		<nav>
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]: We should keep the wrapper block props here
```suggestion
	const blockProps = useBlockProps();

	return (
		<nav { ...blockProps }>
```

Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]: Whoops yup, thanks I'll apply this.

File: plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx [hunk 6]
```diff
@@ -1,35 +1,64 @@
 /**
  * External dependencies
  */
-import { useBlockProps, useInnerBlocksProps } from '@wordpress/block-editor';
+import { InnerBlocks } from '@wordpress/block-editor';
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]: We need to add the import for `useBlockProps` back, I'm sorry I forgot to add this suggestion before.
```suggestion
import { useBlockProps, InnerBlocks } from '@wordpress/block-editor';
```

Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]: Whoops no probs. I'll do this manually in my editor.

Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/blocks/collection-filters/edit.tsx]: Resolved in https://github.com/woocommerce/woocommerce/pull/43216/commits/27e44602e6931ca735c243142b2b9cae483f5295

**Pull request merged**